### DISCUSSION TITLE
Termination with an expression of objectives

### DISCUSSION THREAD

**Discussion by dinohsu1019** [anyoptimization/pymoo#527] [Q&A] [answered]
I am solving my discrete multi-objective problem with NSGA-II, it seems I have no control over when it terminates with a tolerance of say 0.00001, etc., when the result is still unsatisfactory. Is it possible to use an expression of objectivs as termination criteria? 
In the meantime, I am also thinking about (1) larger population and n_offsprings (from 60, 30 to 100, 50) (2) SPX and PM paraters: probability and eta, but not sure how (bigger or smaller?) (3) elitism, is this available?

Please advise, thanks.

**Comment by blankjul** [comment, marked as answer]
Yes you have full control over termination and can write your own logic. I have coded up an example for you below where I am testing if the ideal point is below a specific value. The `MyTermination` overwrites the `_update`  method where you can return a 1 if terminated and a 0 otherwise.


```
import numpy as np

from pymoo.algorithms.moo.nsga2 import NSGA2
from pymoo.core.termination import Termination
from pymoo.optimize import minimize
from pymoo.problems import get_problem
from pymoo.visualization.scatter import Scatter


class Mytermination(Termination):

    def _update(self, algorithm):
        ideal = algorithm.opt.get("F").min(axis=0)
        print(ideal)
        return 1 if np.all(ideal <= 0.1) else 0


problem = get_problem("zdt1")



algorithm = NSGA2(pop_size=100)

res = minimize(problem,
               algorithm,
               termination=Mytermination(),
               seed=1,
               verbose=False)

plot = Scatter()
plot.add(problem.pareto_front(), plot_type="line", color="black", alpha=0.7)
plot.add(res.F, facecolor="none", edgecolor="red")
plot.show()
```

I hope this helps.

**Reply by dinohsu1019** [reply]
A couple of additional points:
1. In cases when the optimal solution (based on my custom criteria of selecting one from nds) can have no changes over many iterations: 
https://github.com/anyoptimization/pymoo/discussions/528
I will modify the _update method to exit if a threashold number of iterations have passed.

2. However, in your sample code, zero is returned if not time for exit yet, I wonder if this can cause the following error: 
https://github.com/anyoptimization/pymoo/discussions/533

If this is the case, I need to return the same return value as the original _update method, so is this correct? (with "else super._update()")

```
# 2023-12-22 新增：N_GEN_WITH_SAME_OPT 當若干代都得到相同的最佳解時即結束演算法
#            最佳解目前使用 get_TNeagle_opt() 從每一代的nds得出
class MyTermination(DefaultMultiObjectiveTermination):

    def __init__(self, **kwargs):
        assert N_GEN_WITH_SAME_OPT>1 and N_GEN_WITH_SAME_OPT<=N_MAX_GEN
        super().__init__(
            ftol = FTOL,
            period = PERIOD,
            n_max_gen = N_MAX_GEN,
            n_max_evals = N_MAX_EVALS,
            **kwargs
        )
        self.n_gen_with_same_opt = N_GEN_WITH_SAME_OPT 
            # if n_gen with the same optimal solution, exit algorithm
        self.opt_objs = []

    def _update(self, algorithm):
        opt_F, opt_X, opt_index = get_TNeagle_opt(algorithm.opt.get("F"),algorithm.opt.get("X"),IDEAL_MDD)
        self.opt_objs.append(tuple(opt_F))
        opt_solution_found = (len(self.opt_objs) >= self.n_gen_with_same_opt and \
            len(set(self.opt_objs[-self.n_gen_with_same_opt:])) == 1) 
        return 1 if opt_solution_found else super()._update(algorithm)
.........................
termination = MyTermination()

```

I am not into object-orietned programming too much, so please confirm, especially the last statement: else super()._update(algorithm), (does FTOL still work?) thanks.
